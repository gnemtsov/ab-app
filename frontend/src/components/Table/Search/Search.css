.Search {
	width: 100%;
	padding: 0 1rem 0 0;
	
	display: grid;
	grid-template-columns: 5fr 1fr;
	grid-template-rows: auto auto;
	grid-template-areas: "input-label option-label" "input option";
}
