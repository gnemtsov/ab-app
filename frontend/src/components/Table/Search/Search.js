import React, { Component } from 'react';
import memoize from 'memoize-one';

import FormElement from '../../Form/FormElement/FormElement';

import classes from './Search.css';

// TODO: search by multiple options
class Search extends Component {
	state = {
		search: "",
		selected: null, // {title, name}
		options: {} // {title0: name0, title1: name1, ...}
	}
	
	constructor(props) {
		super(props);
		
		// Build new options from columns
		const options = {};
		props.cols.forEach( col => options[col.title] = col.name);
		
		const key = Object.keys(options)[0];
		const selected = {
			title: key,
			name: options[key]
		};
		
		this.state = {
			...this.state,
			options,
			selected
		};
	}
	
	buildFilter = memoize(
		(search, option) => {
			if (search === "" || !option) {
				return null;
			}
			
			const filter = {};
			filter[option.name] = search;
			
			return filter;
		}
	);
	
	componentDidUpdate(prevProps, prevState) {
		if (prevState !== this.state) {
			this.props.onSetFilter(this.buildFilter(this.state.search, this.state.selected));
		}
	}
	
	render() {
		return (
			<div className={classes.Search}>
				<FormElement
					id={this.props.id + '_search'}
					key={this.props.id + '_search'}
					inputChanged={data => this.setState({search: data.target.value})}
					message={''}
					placeholder={'search'}
					label={'search'}
					type={'String'}
					labelStyle={{
						gridArea: 'input-label',
						leftText: true
					}}
					bodyStyle={{
						gridArea: 'input'
					}} />
				<FormElement
					id={this.props.id + '_select'}
					key={this.props.id + '_select'}
					inputChanged={data => {
						this.setState({selected: {title: data.target.value, name: this.state.options[data.target.value]}})
					}}
					message={''}
					label={'search by'}
					allowedValues={Object.keys(this.state.options)}
					type={'String'}
					labelStyle={{
						gridArea: 'option-label',
						leftText: true
					}}
					bodyStyle={{
						gridArea: 'option'
					}} />
			</div>
		);
	}
}

export default Search;
