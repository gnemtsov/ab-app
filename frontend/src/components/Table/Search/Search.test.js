import React from 'react';

import { configure, render, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Search from './Search';

configure({ adapter: new Adapter() });

describe('Search', () => {
	it('Basic render test', () => {
		const cols = [
			{ name: 'name', title: 'Pet name' },
			{ name: 'class', title: 'Animal class' },
			{ name: 'age', title: 'Age' },
			{ name: 'gender', title: 'Gender' }
		];
		
		const wrapper = shallow (
	        <Search
				id = {"search"}
				cols={cols}
				 />
		);
		
		expect(wrapper).toMatchSnapshot();
	});
	
	it('Options', () => {
		const cols = [
			{ name: 'name', title: 'Pet name' },
			{ name: 'class', title: 'Animal class' },
			{ name: 'age', title: 'Age' },
			{ name: 'gender', title: 'Gender' }
		];
		
		const wrapper = shallow (
	        <Search
				id = {"search"}
				cols={cols}
				 />
		);
		
		expect(wrapper.state('options')).toEqual({
			'Pet name': 'name',
			'Animal class': 'class',
			'Age': 'age',
			'Gender': 'gender'
		});
	});
	
	it('Change search input', () => {
		const cols = [
			{ name: 'name', title: 'Pet name' },
			{ name: 'class', title: 'Animal class' },
			{ name: 'age', title: 'Age' },
			{ name: 'gender', title: 'Gender' }
		];
		
		const mockOnSetFilter = jest.fn(filter => {});
		
		const wrapper = shallow (
	        <Search
				id = {'search'}
				cols={cols}
				onSetFilter={mockOnSetFilter}
				 />
		);
		const formElements = wrapper.find('FormElement');
		formElements.get(0).props.inputChanged({
			target: {
				value: '14'
			}
		});
		formElements.get(1).props.inputChanged({
			target: {
				value: 'Age'
			}
		});
		
		expect(wrapper.state('search')).toBe('14');
		expect(wrapper.state('selected')).toEqual({title: 'Age', name: 'age'});
		
		expect(mockOnSetFilter.mock.calls[1][0]).toEqual({
			'age': '14'
		});
	});
	
	it('buildFilter', () => {
		const wrapper = shallow (
	        <Search
				id = {'search'}
				cols={[]}
				 />
		);
		
		expect(wrapper.instance().buildFilter('Molly', {title: 'Pet name', name: 'name'})).toEqual({
			name: 'Molly'
		});
	})
});
