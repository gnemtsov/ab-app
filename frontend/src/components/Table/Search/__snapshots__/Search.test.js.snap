// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Basic render test 1`] = `
<div
  className="Search"
>
  <FormElement
    bodyStyle={
      Object {
        "gridArea": "input",
      }
    }
    id="search_search"
    inputChanged={[Function]}
    key="search_search"
    label="search"
    labelStyle={
      Object {
        "gridArea": "input-label",
        "leftText": true,
      }
    }
    message=""
    placeholder="search"
    type="String"
  />
  <FormElement
    allowedValues={
      Array [
        "Pet name",
        "Animal class",
        "Age",
        "Gender",
      ]
    }
    bodyStyle={
      Object {
        "gridArea": "option",
      }
    }
    id="search_select"
    inputChanged={[Function]}
    key="search_select"
    label="search by"
    labelStyle={
      Object {
        "gridArea": "option-label",
        "leftText": true,
      }
    }
    message=""
    type="String"
  />
</div>
`;
